/*
 *  Copyright (c) 2014, Lukas Tenbrink.
 *  * http://lukas.axxim.net
 */

package ivorius.psychedelicraft.block;

import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.function.Supplier;
import java.util.stream.Stream;

import org.jetbrains.annotations.Nullable;

import com.google.common.base.Suppliers;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;

import net.minecraft.block.*;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.*;
import net.minecraft.loot.LootTable;
import net.minecraft.loot.context.*;
import net.minecraft.registry.Registries;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.sound.BlockSoundGroup;
import net.minecraft.sound.SoundEvents;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.*;
import net.minecraft.util.*;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Vec3d;
import net.minecraft.util.math.random.Random;
import net.minecraft.world.*;
import net.minecraft.world.block.OrientationHelper;
import net.minecraft.world.block.WireOrientation;

public class BurdenedLatticeBlock extends LatticeBlock implements Fertilizable {
    public static final MapCodec<BurdenedLatticeBlock> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            Codec.BOOL.fieldOf("spreads").forGetter(block -> block.spreads),
            Registries.BLOCK.getCodec().fieldOf("stem").forGetter(block -> block.stem),
            Codec.INT.fieldOf("shearedAge").forGetter(block -> block.shearedAge),
            AbstractBlock.createSettingsCodec()
    ).apply(instance, BurdenedLatticeBlock::new));

    public static final IntProperty AGE = Properties.AGE_3;
    public static final BooleanProperty PERSISTENT = Properties.PERSISTENT;
    public static final int MAX_AGE = 3;

    private boolean spreads;

    @Nullable
    private final Block stem;

    private final int shearedAge;

    private final Supplier<Optional<RegistryKey<LootTable>>> farmingLootTableKey = Suppliers.memoize(() -> {
        return getLootTableKey().map(key -> RegistryKey.of(RegistryKeys.LOOT_TABLE, key.getValue().withSuffixedPath("_farming")));
    });

    public BurdenedLatticeBlock(boolean spreads, @Nullable Block stem, int shearedAge, Settings settings) {
        super(settings);
        this.spreads = spreads;
        this.stem = stem;

        this.shearedAge = shearedAge;
        setDefaultState(getDefaultState().with(AGE, 0).with(PERSISTENT, false));
    }

    @Override
    protected MapCodec<? extends BurdenedLatticeBlock> getCodec() {
        return CODEC;
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        super.appendProperties(builder);
        builder.add(AGE, PERSISTENT);
    }

    @Override
    protected BlockSoundGroup getSoundGroup(BlockState state) {
        if (state.get(AGE) > 0) {
            return BlockSoundGroup.GRASS;
        }
        return super.getSoundGroup(state);
    }

    @Override
    protected List<ItemStack> getDroppedStacks(BlockState state, LootWorldContext.Builder builder) {
        ItemStack tool = builder.getOptional(LootContextParameters.TOOL);
        if (tool != null && !tool.isEmpty() && builder.getWorld().getRegistryManager().getOrThrow(RegistryKeys.ENCHANTMENT)
                    .getEntry(Enchantments.SILK_TOUCH.getValue())
                    .filter(ench -> EnchantmentHelper.getEnchantments(tool).getLevel(ench) > 0)
                    .isPresent()) {
            ItemStack drop = asItem().getDefaultStack();
            drop.setDamage(state.get(AGE));
            return List.of(drop);
        }
        return super.getDroppedStacks(state, builder);
    }

    @Override
    protected ActionResult onUseWithItem(ItemStack stack, BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hit) {
        if (stack.isOf(Items.SHEARS)) {

            if (state.get(AGE) < MAX_AGE || state.get(PERSISTENT)) {
                return ActionResult.FAIL;
            }

            world.playSoundFromEntity(null, player, SoundEvents.ENTITY_SHEEP_SHEAR, player.getSoundCategory(), 1, 1);

            if (!world.isClient) {
                getFarmingLootTableKey()
                    .map(world.getServer().getReloadableRegistries()::getLootTable)
                    .ifPresent(table -> table.generateLoot(new LootWorldContext.Builder((ServerWorld)world)
                            .add(LootContextParameters.ORIGIN, Vec3d.ofCenter(pos))
                            .add(LootContextParameters.TOOL, player.getStackInHand(hand))
                            .add(LootContextParameters.BLOCK_STATE, state)
                            .addOptional(LootContextParameters.BLOCK_ENTITY, world.getBlockEntity(pos))
                            .build(LootContextTypes.BLOCK)).forEach(s -> Block.dropStack(world, pos, s)));
                world.setBlockState(pos, state.with(AGE, shearedAge));
            }
            if (!player.isCreative()) {
                player.getStackInHand(hand).damage(1, player, hand == Hand.MAIN_HAND ? EquipmentSlot.MAINHAND : EquipmentSlot.OFFHAND);
            }
            return ActionResult.SUCCESS;
        }

        return ActionResult.PASS_TO_DEFAULT_BLOCK_ACTION;
    }

    public final Optional<RegistryKey<LootTable>> getFarmingLootTableKey() {
        return farmingLootTableKey.get();
    }

    @Override
    protected void randomTick(BlockState state, ServerWorld world, BlockPos pos, Random random) {
        if (state.get(PERSISTENT)) {
            return;
        }

        if (checkConnectivity(world, state, pos)) {
            if (world.getLightLevel(pos) >= 9 && random.nextInt(35) == 0) {
                if (state.get(AGE) < MAX_AGE) {
                    world.setBlockState(pos, state.cycle(AGE), Block.NOTIFY_ALL);
                    WireOrientation wireOrientation = OrientationHelper.getEmissionOrientation(
                        world, null, Direction.UP
                    );
                    world.updateNeighborsAlways(pos, this, wireOrientation);
                }

                if (state.get(AGE) > 0) {
                    trySpread(state, world, pos);
                }
            }
        } else if (state.get(AGE) > 0) {
            world.setBlockState(pos, state.with(AGE, state.get(AGE) - 1), Block.NOTIFY_ALL);
        }
    }

    @Override
    public boolean isFertilizable(WorldView world, BlockPos pos, BlockState state) {
        return (state.get(AGE) < MAX_AGE || canSpread(state, world, pos)) && !state.get(PERSISTENT);
    }

    @Override
    public boolean canGrow(World world, Random random, BlockPos pos, BlockState state) {
        return true;
    }

    @Override
    public void grow(ServerWorld world, Random random, BlockPos pos, BlockState state) {
        if (state.get(AGE) < MAX_AGE) {
            world.setBlockState(pos, state.with(AGE, MAX_AGE), Block.NOTIFY_ALL);
        } else {
            trySpread(state, world, pos);
        }
    }

    @Override
    public ItemStack getPickStack(WorldView world, BlockPos pos, BlockState state, boolean includeData) {
        ItemStack stack = super.getPickStack(world, pos, state, includeData);
        if (includeData) {
            stack.setDamage(state.get(AGE));
        }
        return stack;
    }

    @Override
    public BlockState getPlacementState(ItemPlacementContext ctx) {
        return super.getPlacementState(ctx)
                .with(AGE, ctx.getStack().getDamage() % (MAX_AGE + 1))
                .with(PERSISTENT, true);
    }

    private void trySpread(BlockState state, World world, BlockPos pos) {
        if (!spreads) {
            return;
        }
        visitNeighbours(state, pos)
            .filter(mPos -> world.getBlockState(mPos).isOf(PSBlocks.LATTICE))
            .findFirst()
            .ifPresent(mPos -> world.setBlockState(mPos, copyStateProperties(getDefaultState(), world.getBlockState(mPos))));
    }

    private boolean canSpread(BlockState state, WorldView world, BlockPos pos) {
        return spreads && visitNeighbours(state, pos).anyMatch(mPos -> world.getBlockState(mPos).isOf(PSBlocks.LATTICE));
    }

    public boolean checkConnectivity(WorldView world, BlockState state, BlockPos pos) {
        if (stem == null) {
            return true;
        }

        Set<BlockPos> visitedPositions = new HashSet<>();
        visitedPositions.add(pos.toImmutable());
        return checkConnectivity(world, state, pos, visitedPositions, 20);
    }

    private boolean checkConnectivity(WorldView world, BlockState state, BlockPos pos, Set<BlockPos> visitedPositions, int maxDepth) {

        if (state.isOf(stem)) {
            return true;
        }

        if (!state.isOf(this)) {
            return false;
        }

        if (getFreeConnections(state).anyMatch(facing -> world.getBlockState(pos.offset(facing)).isOf(stem))) {
            return true;
        }

        if (maxDepth > 1) {
            return visitNeighbours(state, pos)
                    .filter(visitedPositions::add)
                    .anyMatch(neighbourPos -> checkConnectivity(world, world.getBlockState(neighbourPos), neighbourPos, visitedPositions, maxDepth - 1));
        }

        return false;
    }

    private static Stream<BlockPos> visitNeighbours(BlockState state, BlockPos pos) {
        return getConnections(state).flatMap(d -> {
            BlockPos onLevel = pos.offset(d);
            return Stream.of(onLevel.up(), onLevel, onLevel.down());
        });
    }
}
